scripts/skills/effects/rf_soul_link_effect.nut

```diff
@@ -0,0 +1,122 @@
+		this.m.IsRemovedAfterBattle = true;
+	}
+
+	function getName()
+	{
+		if (this.hasLink()) return (this.m.Name + " (Active)");
+		return this.m.Name;
+	}
+
+	function getTooltip()
+	{
+		local tooltip = this.skill.getTooltip();
+
+		tooltip.push({
+			id = 10,
+			type = "text",
+			icon = "ui/icons/health.png",
+			text = ::MSU.Text.colorGreen((this.m.TransferedPart * 100) + "%") + " of Hitpoint damage taken is transferred to a random adjacent ally that does not have Soul Link"
+		});
+
+		return tooltip;
+	}
+
+	function onDamageReceived( _attacker, _damageHitpoints, _damageArmor )
+	{
+		if (_damageHitpoints == 0) return;
+		if (!this.hasLink()) return;
+
+		local transferedDamageMult = (1.0 / (1.0 - this.m.TransferedPart)) - 1.0;
+		this.m.TransferInfo = {
+			Attacker = _attacker.weakref(),
+			Damage = ::Math.ceil(transferedDamageMult * _damageHitpoints * this.m.TransferedMult)
+		}
+	}
+
+	function onAfterDamageReceived()	// Transferred damage should display after the initial damage in the log
+	{
+		if (this.m.TransferInfo == null) return;
+		this.transferDamage(this.m.TransferInfo.Attacker, this.m.TransferInfo.Damage);
+		this.m.TransferInfo = null;
+	}
+
+	function onDeath(_killer, _skill, tile, _fatalityType)	// Transferred damage should display after the death in the log
+	{
+		if (this.m.TransferInfo == null) return;
+		this.transferDamage(this.m.TransferInfo.Attacker, this.m.TransferInfo.Damage);
+		this.m.TransferInfo = null;
+	}
+
+	function onUpdate( _properties )
+	{
+		if (this.hasLink()) _properties.DamageReceivedTotalMult *= (1.0 - this.m.TransferedPart);
+	}
+
+	function hasLink()
+	{
+		if (this.getLink() == null) return false;
+		return true;
+	}
+
+	function getLink()
+	{
+		local candidates = [];
+		local actor = this.getContainer().getActor();
+		local myTile = actor.getTile();
+		for( local i = 0; i < 6; i = ++i )
+		{
+			if (myTile.hasNextTile(i) == false) continue;
+
+			local nextTile = myTile.getNextTile(i);
+			if (!nextTile.IsOccupiedByActor) continue;
+
+			local nextEntity = nextTile.getEntity();
+			if (!nextEntity.isAlliedWith(actor)) continue;
+			if (nextEntity.getSkills().hasSkill(this.m.ID)) continue;
+
+			candidates.push(nextEntity);
+		}
+		if (candidates.len() == 0) return null;
+		return candidates[::Math.rand(0, candidates.len() - 1)];
+	}
+
+	function transferDamage( _attacker, _damage )
+	{
+		local target = this.getLink();
+		if (this.m.SoundOnUse.len() != 0)
+		{
+			::Sound.play(this.m.SoundOnUse[::Math.rand(0, this.m.SoundOnUse.len() - 1)], ::Const.Sound.Volume.RacialEffect, target.getPos());
+		}
+
+		local hitInfo = clone ::Const.Tactical.HitInfo;
+		hitInfo.DamageRegular = _damage;
+		hitInfo.DamageDirect = 1.0;
+		hitInfo.BodyPart = ::Const.BodyPart.Body;
+		hitInfo.BodyDamageMult = 1.0;
+		hitInfo.FatalityChanceMult = 0.0;
+		target.onDamageReceived(_attacker, this, hitInfo);	// Forwarding the attacker skill instead of 'this' would be nice but we don't have access to that one
```

Comment:
You can store the `_skill` of the attacker in the `this.m.TransferInfo` from `onBeforeDamageReceived`. However, I also see arguments for dealing the damage to the linked target from this character and not the `_attacker`. In that case we should be passing `this.getContainer().getActor()` instead of `_attacker`.

For reference, vanilla `hex_slave_effect` applies the Hex damage from the hexed character to himself.

Reply:
I don't want experience/loot to be voided when we kill a body guard via the soul link damage.
So I assumed passing the attacker here will make sure that the player doesn't lose out on any loot

Reply:
That's a good point. Yeah then let's pass the `_attacker`. But then please store the skill used and pass that too.

Reply:
Ok will do that. But I don't know whether that would be the best way.
Is it really that attacker skill or the soul link that caused the kill? - Somewhat philosophical question

Reply:
Perhaps passing `this` _is_ better because then the KilledString and other effects related to the death of the linked character are handled properly e.g. we wouldn't want Overhead Strike on a Necromancer to decapitate a linked zombie.